service-config/src/main/java/com/palantir/config/service/ServiceDiscoveryConfiguration.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.config.service;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.google.common.base.Optional;
+import com.google.common.base.Strings;
+import com.palantir.remoting.ssl.SslConfiguration;
+import com.palantir.tokens.auth.BearerToken;
+import java.util.Map;
+import org.immutables.value.Value.Immutable;
+import org.immutables.value.Value.Style;
+
+/**
+ * Configuration class that contains a map of {@code serviceName}s and their respective {@link ServiceConfiguration}s
+ * and provides utility methods for apps to build service clients.
+ */
+@Immutable
+@JsonDeserialize(as = ImmutableServiceDiscoveryConfiguration.class)
+@Style(visibility = Style.ImplementationVisibility.PACKAGE)
+public abstract class ServiceDiscoveryConfiguration {
+
+    @JsonProperty("apiToken")
+    public abstract Optional<BearerToken> defaultApiToken();
+
+    @JsonProperty("security")
+    public abstract Optional<SslConfiguration> defaultSecurity();
+
+    public abstract Map<String, ServiceConfiguration> services();
+
+    /**
+     * Uses the provided service name as the key to retrieve the API token for the respective
+     * {@link ServiceConfiguration}. If the API token is not defined in the {@link ServiceConfiguration}, the default
+     * API token will be returned.
+     *
+     * @param serviceName
+     *        the name of the service to be used to retrieve the respective {@link ServiceConfiguration}
+     * @returns the API token for a service based on the input service name
+     */
+    public final Optional<BearerToken> getApiToken(String serviceName) {
```

Comment:
It was discussed in the quip doc linked below that we want to have service specific token if they are present, but default to the root level configurations if not. Hence bundling the logic here to make sure the behavior is consistent across consuming apps